### PR TITLE
Added KanrenRelationSub for distributive rewrites

### LINKED ISSUE #606: Missing rational equation simplifications
We're missing some simplifications for rational equations.  For example:

``` python
import aesara
import aesara.tensor as at


eta_at = at.scalar("eta")
kappa_at = at.scalar("kappa")

# This is equivalent to `1 / kappa_at`
graph_at = eta_at / kappa_at + (1 - eta_at) / kappa_at

graph_fn = aesara.function([eta_at, kappa_at], graph_at)

# The "optimized" graph still contains the unnecessary `eta` parameter:
aesara.dprint(graph_fn.maker.fgraph)
# Elemwise{Composite{((i0 / i1) + ((i2 - i0) / i1))}} [id A] ''   0
#  |eta [id B]
#  |kappa [id C]
#  |TensorConstant{1.0} [id D]
```

Without these simplifications, issues can arise when people use parameterizations like `alpha = eta / kappa`, `beta = (1 - eta) / kappa` in&mdash;for example&mdash;beta-binomial models.

### PR DESCRIPTION
**Pull request by kc611**
This PR adds the following optimizations
- [x] x*a + y*a + z*a -> (x + y + z)*a
- [x]  x/a + y/a + z/a -> (x + y + z)/a
- [x] x*a - y*a -> (x - y)*a
- [x] x/a - y/a -> (x - y)/a

Resolves #606

The graph now returns
```python
import aesara
import aesara.tensor as at

eta_at = at.scalar("eta")
kappa_at = at.scalar("kappa")

graph_at = eta_at / kappa_at + (1 - eta_at) / kappa_at
graph_fn = aesara.function([eta_at, kappa_at], graph_at)

aesara.dprint(graph_fn.maker.fgraph)
# Elemwise{reciprocal,no_inplace} [id A] ''   0
#  |kappa [id B]
```
Gist elaborating the implementation: 
https://gist.github.com/kc611/b33e45ed2086597ed9c9df4f387c84b0

### CODE REVIEW

File: aesara/configdefaults.py [hunk 1]
```diff
@@ -744,6 +744,15 @@ def add_compile_configvars():
         in_c_key=False,
     )
 
+    config.add(
+        "fastmath_opts",
```
Comment (kc611) [inline aesara/configdefaults.py:748]: Added this configuration which will make the rewrite optional.

Comment (brandonwillard) [reply aesara/configdefaults.py:748]: Instead, don't make them canonicalizations (i.e. don't use `register_canonicalize`) and register them manually to a different optimization DB.  As it currently stands, `local_add_sub_collector` is being called unnecessarily far too often.

In general, optimization selection and filtering is accomplished using the optimizations DBs and tags.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

**Comment by kc611** [comment]
Alright, so I registered onto a `SequenceDB` but when I do:

```python
import aesara
import aesara.tensor as at
from aesara.graph.optdb import OptimizationQuery
from aesara.tensor.math_opt import fastmath_db
from aesara.graph.fg import FunctionGraph

eta_at = at.scalar("eta")
kappa_at = at.scalar("kappa")

graph_at = eta_at / kappa_at + (1 - eta_at) / kappa_at
graph_fn = FunctionGraph(
        inputs = [eta_at, kappa_at],
        outputs=[graph_at],
    )

fastmath_db.query(OptimizationQuery(include=["basic"])).optimize(graph_fn) #AttributeError: 'FromFunctionLocalOptimizer' object has no attribute 'optimize'

aesara.dprint(graph_fn)
```

Is it not supposed to be used like that ?

**Comment by kc611** [comment]
xa + ya + za -> (x + y + z)a
x/a + y/a + z/a -> (x + y + z)/a
xa - ya -> (x - y)*a
x/a - y/a -> (x - y)/a

Is this particular collection of rewrites subject to #649, should we be separating them out?

**Comment by brandonwillard** [comment]
> Is this particular collection of rewrites subject to #649, should we be separating them out?

Do you mean the `+`/`-` or arity difference?  The `+`/`-` difference is a case of #648.  Otherwise, these rewrites should be able to handle all arities.

**Comment by brandonwillard** [comment]
Now that we have `kanren` support, this would be a good exercise for that.

Using `kanren` would allow us to more succinctly perform all the intermediate computations without the need for potentially (numerically) destabilizing canonicalizations.

We would need to devise a `kanren` relation (well, a *goal constructor*) that uses the properties implemented here (i.e. distributive properties) to "search" for the reductions described in #606.  We would need the resulting goals to succeed only when at least one reduction has been made.  That could be a little tricky to do entirely in miniKanren, though.

Anyway, we can start discussing it here (or somewhere else).

**Comment by kc611** [comment]
I made a kanren version of those distributive optimizations but it ind-of seems like it isn't able to generalize them to an arbitrary number of distributive optimizations. Have a look at https://gist.github.com/kc611/b33e45ed2086597ed9c9df4f387c84b0

**Comment by brandonwillard** [comment]
> I made a kanren version of those distributive optimizations but it ind-of seems like it isn't able to generalize them to an arbitrary number of distributive optimizations. Have a look at https://gist.github.com/kc611/b33e45ed2086597ed9c9df4f387c84b0

Just added a comment in that Gist.

File: aesara/tensor/math_opt.py [hunk 2]
```diff
@@ -3523,6 +3532,36 @@ def local_reciprocal_1_plus_exp(fgraph, node):
...
+        tailo(add_cdr_lv, add_flat_lv),
+        # Add all the `at.mul`ed arguments and set the output
+        conso(at.add, dot_cdr_lv, out_lv),
+        # Apply the `at.mul` to all the flattened `add` arguments
+        mapo(lambda x, y: conso(at.mul, etuple(A_lv, x), y), add_cdr_lv, dot_cdr_lv),
+    )
+
+
+distribute_mul_over_add_opt = KanrenRelationSub(
+    lambda x, y: distribute_mul_over_add(y, x)
```
Comment (kc611) [inline aesara/tensor/math_opt.py:3561]: As of now, I could not figure out how to do the reverse using `kanren` but I guess this works just as well.

Comment (brandonwillard) [reply aesara/tensor/math_opt.py:3561]: Are you referring to the argument swap?

Comment (kc611) [reply aesara/tensor/math_opt.py:3561]: No, I mean I'm yet to figure out how to do xa + ya + z*a -> (x + y + z)*a purely in kanren (without the argument swap). (Though that might not be necessary.)

Comment (brandonwillard) [reply aesara/tensor/math_opt.py:3561]: If the relation encoded by the `kanren` goal `distribute_mul_over_add` is `x*a + y*a + z*a = (x + y + z)*a`, in that exact left-hand/right-hand side order, then `distribute_mul_over_add(l, r)` will unify `l` with `x*a + y*a + z*a` and/or `r` with `(x + y + z)*a`: i.e. the first argument, `l`, will yield "expanded" results, and the second argument, `r`, will yield the "collected" results.

File: tests/tensor/test_math_opt.py [hunk 3]
```diff
@@ -4549,3 +4549,49 @@ def logit_fn(x):
...
+
+class TestDistributiveOpts:
+    x_at = vector("x")
+    y_at = vector("y")
+    a_at = matrix("a")
+
+    @pytest.mark.parametrize(
+        "orig_operation, optimized_operation",
+        [
+            (a_at * x_at + a_at * y_at, a_at * (x_at + y_at)),
```
Comment (ricardoV94) [inline tests/tensor/test_math_opt.py:4562]: Small nitpick, the `_at` suffix makes it harder to read the tests

Comment (brandonwillard) [reply tests/tensor/test_math_opt.py:4562]: They become extremely helpful when you're dealing with multiple variables of different types.  For instance, when using `kanren`, it's good to know that `x_lv` is the logic variable version of some term `x`, and that `x_at` is the Aesara tensor version.

Comment (kc611) [reply tests/tensor/test_math_opt.py:4562]: I'll just add the above said as a comment to make it easier to read/understand later on.

**Comment by kc611** [comment]
The current failing test seems to need a bit of discussion. So it seems that for this particular case of bad `ViewMap` view.
https://github.com/aesara-devs/aesara/blob/e476ebcdd4222758cddd4523cd19248c88fb4dd2/tests/compile/test_debugmode.py#L523-L556

Instead of the error that it's supposed to throw, i.e. `BadViewMap`, it throws the below stated `BadOptimizationError`.

### Stack Trace for the test failure:

```
___________________________________ TestViewMap.test_aliased_outputs_bad ___________________________________

self = <tests.compile.test_debugmode.TestViewMap object at 0x7f13a98d2e50>

    def test_aliased_outputs_bad(self):
        # here the alias between outputs is not ok because destroying one
        # destroys the other, but there's no way to warn aesara about it
        # through the view_map mechanism.
        class CustomOp(Op):
            def make_node(self, a, b):
                c = a.type()
                d = a.type()
                return Apply(self, [a, b], [c, d])
    
            def perform(self, node, inp, out):
                a, b = inp
                c, d = out
                r = a * 1
                c[0] = r[:-1]
                d[0] = r[1:]
    
        custom_op = CustomOp()
    
        x = dvector()
        y = dvector()
        bad_xy0, bad_xy1 = custom_op(x, y)
        out = bad_xy0 * 2 + bad_xy1 * 2
        f = function([x, y], out, mode="DEBUG_MODE")
    
        # with pytest.raises(BadViewMap):
        f([1, 2, 3, 4], [5, 6, 7, 8])

tests/compile/test_debugmode.py:549: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
aesara/compile/function/types.py:975: in __call__
    if output_subset is None
aesara/compile/debugmode.py:2048: in deco
    return f()
aesara/compile/debugmode.py:1973: in f
    order, fgraph.equivalence_tracker.reasons, r_vals
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

order = [CustomOp(<TensorType(float64, vector)>, <TensorType(float64, vector)>), Elemwise{Mul}[(0, 1)](TensorConstant{(1,) of ... 4.0}, CustomOp.0), InplaceDimShuffle{x}(TensorConstant{4.0}), Elemwise{Cast{float64}}(TensorConstant{(1,) of 4}), ...]
reasons = {InplaceDimShuffle{x}.0: [('MergeOptimizer', InplaceDimShuffle{x}.0, "  InplaceDimShuffle{x} [id A] <TensorType(int8, ...pe(int8, scalar)>\n', '  TensorConstant{2} [id B] <TensorType(int8, scalar)>\n')], CustomOp.0: [], CustomOp.1: [], ...}
r_vals = {<TensorType(float64, vector)>: array([1., 2., 3., 4.]), <TensorType(float64, vector)>: array([5., 6., 7., 8.]), TensorConstant{(1,) of 4.0}: array([4.]), TensorConstant{4.0}: array(4.), ...}

    def _find_bad_optimizations0(order, reasons, r_vals):
        """
        Use a simple algorithm to find broken optimizations.
    
        This algorithm is simple to understand, but sometimes when there's
        a problem it identifies the wrong optimization as the culprit.
        The problem stems from the fact that results are not evaluated in
        chronological order (looking at when they were introduced to the
        graph).
    
        """
        # iterate over variables looking for values that don't match the
        # values of the variables they replaced.  This is the sign of a
        # broken optimization.
        for i, node in enumerate(order):
            for new_r in node.outputs:
                for reason, r, old_graph_str, new_graph_str in reasons[new_r]:
                    # check if the value for new_r doesn't match the value for r
                    new_r_val = r_vals[new_r]
                    r_val = r_vals[r]
                    assert r.type == new_r.type
    
                    if hasattr(new_r.tag, "values_eq_approx"):
                        check = new_r.tag.values_eq_approx(r_val, new_r_val)
                    elif hasattr(new_r, "values_eq_approx"):
                        # This way will be deprecated later, but not right now
                        check = new_r.values_eq_approx(r_val, new_r_val)
                    else:
                        check = r.type.values_eq_approx(r_val, new_r_val)
                    if not check:
                        raise BadOptimization(
                            old_r=r,
                            new_r=new_r,
                            old_r_val=r_val,
                            new_r_val=new_r_val,
                            reason=reason,
                            old_graph=old_graph_str,
                            new_graph=new_graph_str,
                        )
E                       aesara.compile.debugmode.BadOptimization: BadOptimization Error 
E                         Variable: id 139722219041168 Elemwise{mul,no_inplace}.0
E                         Op Elemwise{mul,no_inplace}(CustomOp.0, TensorConstant{(1,) of 4})
E                         Value Type: <class 'numpy.ndarray'>
E                         Old Value shape, dtype, strides: (3,) float64 (8,)
E                       
E                         Old Value:  [ 6. 10. 14.]
E                         New Value shape, dtype, strides: (3,) float64 (8,)
E                       
E                         New Value:  [ 4.  8. 12.]
E                         Max Abs Diff:  2.0
E                         Mean Abs Diff:  2.0
E                         Median Abs Diff:  2.0
E                         Std Abs Diff:  0.0
E                         Value at Max Diff:  (4.0, 6.0)
E                         Max Rel Diff:  0.2
E                         Mean Rel Diff:  0.12934472934472935
E                         Median Rel Diff:  0.1111111111111111
E                         Std Rel Diff:  0.051873762629018993
E                         Value at Max Diff:  (4.0, 6.0)
E                       
E                         Reason:  <aesara.graph.kanren.KanrenRelationSub object at 0x7f13d8387450>
E                         Old Graph:
E                         Elemwise{add,no_inplace} [id A] <TensorType(float64, vector)> ''   
E                          |Elemwise{mul,no_inplace} [id B] <TensorType(float64, vector)> ''   
E                          | |CustomOp.0 [id C] <TensorType(float64, vector)> ''   
E                          | | |<TensorType(float64, vector)> [id D] <TensorType(float64, vector)>
E                          | | |<TensorType(float64, vector)> [id E] <TensorType(float64, vector)>
E                          | |InplaceDimShuffle{x} [id F] <TensorType(int8, (True,))> ''   
E                          |   |TensorConstant{2} [id G] <TensorType(int8, scalar)>
E                          |Elemwise{mul,no_inplace} [id H] <TensorType(float64, vector)> ''   
E                            |CustomOp.1 [id C] <TensorType(float64, vector)> ''   
E                            |InplaceDimShuffle{x} [id F] <TensorType(int8, (True,))> ''   
E                       
E                         New Graph:
E                         Elemwise{mul,no_inplace} [id I] <TensorType(float64, vector)> ''   
E                          |CustomOp.0 [id C] <TensorType(float64, vector)> ''   
E                          |TensorConstant{(1,) of 4} [id J] <TensorType(int8, (True,))>
E                       
E                       
E                       Hint: relax the tolerance by setting tensor__cmp_sloppy=1
E                         or even tensor__cmp_sloppy=2 for less-strict comparison

aesara/compile/debugmode.py:734: BadOptimization
```

Now if we change the specific operation which happens in the test i.e. `bad_xy0 * 2 + bad_xy1 * 2` to something which isn't being optimized like `bad_xy0  * bad_xy1` the test passed indicating that the mechanism it's supposed to test is still working. But this exposes a caveat, being that any operation which is represented/constructed in the 'bad' way as described in the test and is optimized by any kind of optimization will throw the wrong kind of error. 

TLDR: I think `BadOptimization` is being reached earlier than where `BadViewMap` is supposed to be thrown.

**Review by brandonwillard — COMMENTED** [review]
You can make the `kanren` goals more flexible with respect to the supported `Op`s, instead of creating goals for each one separately.  For instance, make the `Op` a logic variable `op_lv`, add a goal that confirms the value of `op_lv` is one of the accepted `Op`s, and use `op_lv` in place of the `Op`.

To perform the check in the second step, `conde` can be used (e.g. `conde([eq(op_lv, at.mul)], [eq(op_lv, at.true_div)], ...)`), so can type constraints (e.g. `isinstanceo`).  Just note that if you use something like the examples I just gave, they each have their own limitations (e.g. `conde` testing for `at.mul` will only work for those exact `at.mul` `Op` instances).

File: tests/tensor/test_math_opt.py [hunk 4]
```diff
@@ -4544,3 +4544,53 @@ def logit_fn(x):
...
+    y_at = vector("y")
+    a_at = matrix("a")
+
+    # x_lv, if any, is the logic variable version of some term x,
+    # while x_at is the Aesara tensor version for the same.
+    @pytest.mark.parametrize(
+        "orig_operation, optimized_operation",
+        [
+            (a_at * x_at + a_at * y_at, a_at * (x_at + y_at)),
+            (x_at * a_at + y_at * x_at, (x_at + y_at) * a_at),
```
Comment (kc611) [inline tests/tensor/test_math_opt.py:4560]: So I managed to generalize them for `div` and `mul` but it seems that these rewrites aren't taking the `facts()` into account. For instance this particular case passes but the case above it doesn't. Isn't this supposed to be handled by `fact(commutative, at.mul)` in the rewrite.

Comment (brandonwillard) [reply tests/tensor/test_math_opt.py:4560]: This might have to do with the `eq` goal that's being used at specific steps (even behind the scenes).  There are special `eq_*` goals that do (and don't) take into account the associativity/commutativity (AC) information set by `facts`, and you may need to use those explicitly at certain points.  Just try not to use them when they're not needed; otherwise, the streams resulting from the goals will become very long due to all the permutations induced by AC relations.

**Comment by twiecki** [comment]
This might be the wrong place for this discussion, but there are also rewrites with respect to shapes that can be optimized. For example `b * M * a` where where `a` and `b` are scalars and `M` is a matrix are more efficient to rewrite to `b * a * M` so that there is only a single matrix multiplication, instead of two.

**Comment by kc611** [comment]
If a and b are constants I guess that case would be handled by constant folding. But yeah, we would need a rewrite when they're singular variables if we are to implement something like that.

**Comment by brandonwillard** [comment]
> This might be the wrong place for this discussion, but there are also rewrites with respect to shapes that can be optimized. For example `b * M * a` where where `a` and `b` are scalars and `M` is a matrix are more efficient to rewrite to `b * a * M` so that there is only a single matrix multiplication, instead of two.

Are you talking about transforming `((b * M) * a)` to  `((b * a) * M)` so that there's only one matrix/scalar product?  We have some things like that for sums and products in `aesara.tensor.math_opt`, and the `AlgebraicCanonizer` does some similar things, but it looks like we might not have a rewrite that covers that exact case.

Here's a way you can check these kinds of things:
```python
import aesara

import aesara.tensor as at
from aesara.graph.opt_utils import optimize_graph


a, b = at.scalars("ab")
M = at.matrix("M")

z = (a * M) * b

aesara.dprint(z)
# Elemwise{mul,no_inplace} [id A] ''
#  |Elemwise{mul,no_inplace} [id B] ''
#  | |InplaceDimShuffle{x,x} [id C] ''
#  | | |a [id D]
#  | |M [id E]
#  |InplaceDimShuffle{x,x} [id F] ''
#    |b [id G]

# This will only perform canonicalizations, but others can be added via the
# `include` keyword
z_opt = optimize_graph(z)

aesara.dprint(z_opt)
# Elemwise{mul,no_inplace} [id A] ''
#  |InplaceDimShuffle{x,x} [id B] ''
#  | |a [id C]
#  |M [id D]
#  |InplaceDimShuffle{x,x} [id E] ''
#    |b [id F]


# This will perform all the default optimizations
z_fn = aesara.function([a, b, M], z)

aesara.dprint(z_fn)
# Elemwise{mul,no_inplace} [id A] ''   2
#  |InplaceDimShuffle{x,x} [id B] ''   1
#  | |a [id C]
#  |M [id D]
#  |InplaceDimShuffle{x,x} [id E] ''   0
#    |b [id F]
```

**Comment by ricardoV94** [comment]
Are you guys talking about https://github.com/aesara-devs/aesara/issues/287?

**Comment by brandonwillard** [comment]
> Are you guys talking about #287?

Essentially, yes, but for multiplication.

**Comment by twiecki** [comment]
> Are you talking about transforming ((b * M) * a) to ((b * a) * M) so that there's only one matrix/scalar product?

Yes, exactly.

File: aesara/tensor/math_opt.py [hunk 5]
```diff
@@ -3495,6 +3504,79 @@ def local_reciprocal_1_plus_exp(fgraph, node):
...
+# x_lv, if any, is the logic variable version of some term x,
+# while x_at, if any, is the Aesara tensor version for the same.
+def distribute_over_add(in_lv, out_lv):
+    from cons import cons
+    from kanren import eq, fact, heado, tailo
+    from kanren.assoccomm import eq_assoccomm
+
+    # This does the optimization
+    # 1. (x + y + z) * A  = x * A + y * A + z * A
+    # 2. (x + y + z) / A  = x / A + y / A + z / A
```
Comment (twiecki) [inline aesara/tensor/math_opt.py:3516]: Could this lead to numerical issues?

Comment (kc611) [reply aesara/tensor/math_opt.py:3516]: Ah no, the actual optimization being done here is reverse of that:

```
# Line 3545
distribute_over_add_opt = KanrenRelationSub(lambda x, y: distribute_over_add(y, x))
```

So it combines the additive terms not distributes it. It's just that it's easier to represent it in this way in `kanren` and it works just as well if we implement it the other way round purely in `kanren`.

File: aesara/tensor/math_opt.py [hunk 6]
```diff
@@ -3495,6 +3504,79 @@ def local_reciprocal_1_plus_exp(fgraph, node):
...
+        # Make sure the outer term being is an `add`
+        heado(at.add, add_term_lv),
+        # Flatten the associative pairings of `add` operations
+        assoc_flatten(add_term_lv, add_flat_lv),
+        # Get the flattened `add` arguments
+        tailo(add_cdr_lv, add_flat_lv),
+        # Add all the arguments and set the output
+        eq(cons(at.add, mul_cdr_lv), out_lv),
+        mapo(
+            (lambda x, y: eq_assoccomm(cons(op_lv, etuple(x, A_lv)), y)),
```
Comment (brandonwillard) [inline aesara/tensor/math_opt.py:3538]: In order to handle terms like `(+ A (* A D) ...)`, we could consider extending the `mapo`ed goal so that it handles two cases (e.g. with `conde`/`lany`): the current `(* A D)` product terms and  simply `A`.

For example, something like
```python
lambda x, y: conde([eq_assoccomm(cons(op_lv, etuple(x, A_lv)), y)], [eq(x, one_at), eq(y, A_lv)])
```
where `one_at` is something that unifies with any relevant form of `Constant(1)`.

File: aesara/tensor/math_opt.py [hunk 7]
```diff
@@ -3539,3 +3621,27 @@ def local_reciprocal_1_plus_exp(fgraph, node):
...
+
+
+fastmath = EquilibriumDB()
+
+optdb.register("fastmath", fastmath, 1, "fast_run", "mul")
+
+fastmath.register(
+    "dist_over_add_opt",
+    in2out(distribute_over_add_opt, ignore_newtrees=True),
+    1,
```
Comment (brandonwillard) [inline aesara/tensor/math_opt.py:3633]: `EquillibriumDB` doesn't take a position argument like `SequenceDB`; plus, `SequenceDB` was recently change so that the position argument is now a keyword.  

This numeric value is probably being interpreted as a tag, and it should really raise an exception (looks like we need an issue for that).

File: aesara/tensor/math_opt.py [hunk 8]
```diff
@@ -3539,3 +3621,27 @@ def local_reciprocal_1_plus_exp(fgraph, node):
...
+    "distribute_opts",
+    "fast_run",
+    "mul",
+)
+
+
+fastmath.register(
+    "dist_over_sub_opt",
+    in2out(distribute_over_sub_opt, ignore_newtrees=True),
+    1,
```
Comment (brandonwillard) [inline aesara/tensor/math_opt.py:3643]: See above.

**Review by brandonwillard — COMMENTED** [review]
This PR is turning out to be an amazing example of genuine relational programming (and `kanren`)!

If it's possible to abstract/parameterize the implementation of `distribute_over_add` so that it also covers subtraction, the footprint of these features would be impressively small.

File: aesara/tensor/math_opt.py [hunk 9]
```diff
@@ -3520,6 +3531,71 @@ def local_reciprocal_1_plus_exp(fgraph, node):
...
 
+fact(commutative, aes.mul)
+fact(associative, aes.mul)
+fact(commutative, aes.add)
+fact(associative, aes.add)
+
+
+# x_lv, if any, is the logic variable version of some term x,
+# while x_at, if any, is the Aesara tensor version for the same.
+def distributive_collect(in_lv, out_lv):
```
Comment (twiecki) [inline aesara/tensor/math_opt.py:3542]: I would add a doc-string.

**Comment by twiecki** [comment]
This PR is very powerful in its conciseness and the future it demonstrates. It really brings @brandonwillard's vision to the light. So I think we should turn this into a blog post that describes what's happening, how powerful it is, how unique it is, etc.

**Comment by brandonwillard** [comment]
Looks like the tests are taking too long to finish.  I can't tell if it's GitHub Actions or not, yet.

**Comment by kc611** [comment]
Alright so it seems like there are two separate issues in the failing test over here:

1. Things like `a + a` being optimized into ` a * (1 + 1)`. Now this technically is correct however it's not an optimization so should be fixed, this would need a condition in `kanren` corresponding to `at-least one`. The context in which this would be used is iterating over the contents of the logic variable `cdr_lv` in the `distributive_collect` optimization and checking if at-least one of the values is not equal to `A_lv`
2. The bad-view map error. This seems to be happening simply because the optimization has no way of knowing if there are alias between outputs in that particular case. Last time this happened I remember solving this test case by simply changing the position of the optimization. Not sure what the true fix here should be, though.

**Comment by rlouf** [comment]
I think (1) can be "solved" when what is being discussed [here](https://github.com/aesara-devs/aesara/discussions/1082) is implemented. Constraints like the one you suggested should be kept out of miniKanren goals imho.

Ideally we would apply all the possible rewrites and then use a scoring function to choose the "optimal" graph. We could also adopt a greedy approach a choose the "optimal" rewrite out of all possible rewrites at a given step.

**Comment by brandonwillard** [comment]
> Ideally we would apply all the possible rewrites and then use a scoring function to choose the "optimal" graph. We could also adopt a greedy approach a choose the "optimal" rewrite out of all possible rewrites at a given step.

In this case, we would like to avoid the cost of re-searching the AC "graph space" on each application of these rewrites.

The e-graph data structures mentioned in https://github.com/aesara-devs/aesara/discussions/1082 could help with this, albeit not in a direct way perhaps.  #1165 is a little closer, though, because it sets us up for rewrite caching and the like.  #1165 also helps with https://github.com/aesara-devs/aesara/discussions/1082 and the use of e-graph-like data structures.

**Comment by rlouf** [comment]
Rebased this on `main` and resolved the merge conflicts that appeared after #1054
